src/main/java/uk/gov/hmcts/reform/sscs/services/mapper/CaseDataEventBuilder.java

```diff
@@ -0,0 +1,165 @@
+        postponedEventInferredFromDelta;
+    private final PostponedEventService<Hearing> postponedEventInferredFromCcd;
+
+    @Autowired
+    CaseDataEventBuilder(
+        SearchCcdService searchCcdService, IdamService idamService,
+        PostponedEventService<uk.gov.hmcts.reform.sscs.models.deserialize.gaps2.Hearing>
+            postponedEventInferredFromDelta,
+        PostponedEventService<Hearing> postponedEventInferredFromCcd) {
+        this.searchCcdService = searchCcdService;
+        this.idamService = idamService;
+        this.postponedEventInferredFromDelta = postponedEventInferredFromDelta;
+        this.postponedEventInferredFromCcd = postponedEventInferredFromCcd;
+    }
+
+    List<Events> buildPostponedEvent(AppealCase appealCase) {
+        if (minorStatusIsNotNullAndIsNotEmpty(appealCase.getMinorStatus())) {
+            return appealCase.getMinorStatus().stream()
+                .filter(minorStatus -> areConditionsToCreatePostponedEventMet(minorStatus.getStatusId(), appealCase))
+                .map(minorStatus -> buildNewPostponedEvent(minorStatus.getDateSet()))
+                .distinct()
+                .collect(Collectors.toList());
+        }
+        return Collections.emptyList();
+    }
+
+    private boolean areConditionsToCreatePostponedEventMet(String statusId, AppealCase appealCase) {
+        if (minorStatusIdIs27AndThereIsOnlyOnePostponementRequestGranted(statusId, appealCase)) {
+            return true;
+        }
+
+        if (minorStatusIdIs27AndMoreThanOnePostponementRequest(statusId, appealCase)) {
+            if (postponedEventInferredFromDelta.matchToHearingId(appealCase.getPostponementRequests(),
+                appealCase.getHearing())) {
+                return true;
+            }
+            List<Hearing> hearingList = retrieveHearingsFromCaseInCcd(appealCase);
+            return postponedEventInferredFromCcd.matchToHearingId(appealCase.getPostponementRequests(), hearingList);
+        }
+        return false;
+    }
+
+    @Retryable
+    private List<Hearing> retrieveHearingsFromCaseInCcd(AppealCase appealCase) {
+        IdamTokens idamTokens = IdamTokens.builder()
+            .idamOauth2Token(idamService.getIdamOauth2Token())
+            .authenticationService(idamService.generateServiceAuthorization())
+            .build();
+        List<CaseDetails> caseDetailsList = searchCcdService.findCaseByCaseRef(appealCase.getAppealCaseRefNum(),
+            idamTokens);
+        CaseData caseData = CcdUtil.getCaseData(caseDetailsList.get(0).getData());
+        return caseData.getHearings();
+    }
+
+    private boolean minorStatusIdIs27AndMoreThanOnePostponementRequest(String statusId, AppealCase appealCase) {
+        return "27".equals(statusId) && appealCase.getPostponementRequests() != null
+            && !appealCase.getPostponementRequests().isEmpty() && appealCase.getPostponementRequests().size() > 1;
+    }
```

Comment:
The last check is **appealCase.getPostponementRequests().size()**  is unnecessary as the previous one  **!appealCase.getPostponementRequests().isEmpty()** does the same.